jpa/org.eclipse.persistence.jpa/src/main/java/org/eclipse/persistence/jpa/metadata/FileBasedProjectCacheFilter.java

```diff
@@ -0,0 +1,127 @@
+ * or the Eclipse Distribution License v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+// Contributors:
+//     08/02/2022-4.0.0 Tomas Kraus - 1613: FileBasedProjectCache retrieveProject contains readObject with no data filtering
+package org.eclipse.persistence.jpa.metadata;
+
+import java.io.ObjectInputFilter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.rmi.Remote;
+
+import org.eclipse.persistence.Version;
+import org.eclipse.persistence.logging.SessionLog;
+import org.eclipse.persistence.sessions.Project;
+
+/**
+ * Verify content of FileBasedProjectCache being read with ObjectInputStream.
+ */
+class FileBasedProjectCacheFilter implements ObjectInputFilter {
+
+    private final SessionLog log;
+
+    /**
+     * Creates an instance of ObjectInputStream data of FileBasedProjectCache verifier.
+     */
+    FileBasedProjectCacheFilter(SessionLog log) {
+        this.log = log;
+        //log.setLevel(SessionLog.FINEST);
+    }
+
+    // ObjectInputStream data verification entry point.
+    // This method is being called for every object instance being deserialized.
+    @Override
+    public Status checkInput(FilterInfo info) {
+        try {
+            // Skip checks when decision was already made by higher level filter.
+            ObjectInputFilter serialFilter = ObjectInputFilter.Config.getSerialFilter();
+            if (serialFilter != null) {
+                ObjectInputFilter.Status status = serialFilter.checkInput(info);
+                if (status != ObjectInputFilter.Status.UNDECIDED) {
+                    // The process-wide filter overrides this filter
+                    return status;
+                }
+            }
+            // Check current object instance.
+            final Class<?> serialClass = info.serialClass();
+            if (serialClass != null) {
+                log.log(SessionLog.FINE, String.format(
+                        "FileBasedProjectCacheFilter: processing level %d class %s",
+                        info.depth(), serialClass.getName()));
+                // Only Project class instance is allowed on top level of the structure
+                // and class must also come from known .jar file name.
+                if (info.depth() == 1L) {
+                    if (serialClass != Project.class || !isEclipseLinkFileJar(serialClass, log)) {
+                        log.log(SessionLog.WARNING, String.format(
+                                "FileBasedProjectCacheFilter: rejected illegal top level FileBasedProjectCache class: %s",
+                                serialClass.getName()));
+                        return Status.REJECTED;
+                    }
+                }
+                // Remote interface must be rejected
+                if (info.serialClass() != null && Remote.class.isAssignableFrom(info.serialClass())) {
```

Comment:
not sure this is correct assumption, can you double checkit via [jpa.test.remote](https://github.com/eclipse-ee4j/eclipselink/tree/master/jpa/eclipselink.jpa.testapps/jpa.test.remote) test app? Two tests there are currently expected to fail but it should give an idea what can or cannot be send via RMI..

Reply:
This piece of code was just taken from JDK 9 javadoc as a recomended check - see https://docs.oracle.com/javase/9/docs/api/java/io/ObjectInputFilter.html
I can remove it if you think it can break something.

Reply:
Removed. There is check whether `info.serialClass() == Project.class` which is not implementing Remote. It should be sufficient.